// @chijioke/graphql-client/dist/index.cjs.js
'use strict';

// src/auth.ts
var toAsyncTokenProvider = (p) => {
  if (!p) return async () => null;
  return async () => {
    try {
      const res = p();
      return res instanceof Promise ? await res : res;
    } catch (e) {
      return null;
    }
  };
};
var toAsyncHeadersFactory = (f) => {
  if (!f) return async () => ({});
  return async () => {
    var _a;
    try {
      const res = f();
      return (_a = res instanceof Promise ? await res : res) != null ? _a : {};
    } catch (e) {
      return {};
    }
  };
};

// src/errors.ts
var SdkError = class extends Error {
  constructor(message, code, info) {
    super(message);
    this.name = "SdkError";
    this.code = code;
    this.info = info;
  }
};
var NetworkError = class extends SdkError {
  constructor(message = "Network error", info) {
    super(message, "NETWORK_ERROR", info);
  }
};
var AuthenticationError = class extends SdkError {
  constructor(message = "Authentication failed", info) {
    super(message, "AUTH_ERROR", info);
  }
};

// src/middlewares/cache.ts
var createCacheMiddleware = (opts) => {
  var _a, _b, _c;
  const ttl = (_a = opts == null ? void 0 : opts.ttlMs) != null ? _a : 3e4;
  const keyFn = (_b = opts == null ? void 0 : opts.keyFn) != null ? _b : ((q, v) => JSON.stringify({ q, v }));
  const maxEntries = (_c = opts == null ? void 0 : opts.maxEntries) != null ? _c : 100;
  const store = /* @__PURE__ */ new Map();
  const middleware = async (ctx, resCtx, next) => {
    const { skipCache } = ctx.cacheOptions || {};
    if (!skipCache && ctx.request.query.trim().toLowerCase().startsWith("query")) {
      const key = keyFn(ctx.request.query, ctx.request.variables);
      const now = Date.now();
      const hit = store.get(key);
      if (hit && hit.expiry > now) {
        resCtx.response = hit.value;
        resCtx.status = 200;
        return;
      }
      await next();
      if (resCtx.response && !resCtx.error) {
        store.set(key, { value: resCtx.response, expiry: now + ttl });
        if (store.size > maxEntries) {
          const oldestKey = store.keys().next().value;
          if (oldestKey) store.delete(oldestKey);
        }
      }
    } else {
      await next();
    }
  };
  const cacheMw = middleware;
  cacheMw.clearCache = () => {
    store.clear();
  };
  return cacheMw;
};

// src/middlewares/retry.ts
var retryMiddleware = (opts) => {
  var _a, _b;
  const retries = (_a = opts == null ? void 0 : opts.retries) != null ? _a : 2;
  const delayMs = (_b = opts == null ? void 0 : opts.delayMs) != null ? _b : 150;
  return async (ctx, resCtx, next) => {
    let attempt = 0;
    while (attempt <= retries) {
      await next();
      if (!resCtx.error) return;
      attempt++;
      if (attempt > retries) return;
      await new Promise((r) => setTimeout(r, delayMs * attempt));
    }
  };
};

// src/middlewares/middleware.ts
var compose = (middlewares) => async (ctx, resCtx) => {
  let idx = -1;
  const dispatch = async (i) => {
    if (i <= idx) throw new Error("next() called multiple times");
    idx = i;
    const fn = middlewares[i];
    if (!fn) return;
    await fn(ctx, resCtx, () => dispatch(i + 1));
  };
  await dispatch(0);
};

// src/transport.ts
var createTransport = (opts) => {
  var _a;
  const fetchImpl = (_a = opts == null ? void 0 : opts.fetchImpl) != null ? _a : typeof fetch !== "undefined" ? fetch.bind(globalThis) : void 0;
  if (!fetchImpl) {
    throw new Error(
      "No fetch implementation available. Provide fetchImpl in TransportOptions or polyfill fetch."
    );
  }
  return async (ctx, resCtx, next) => {
    const body = JSON.stringify({ query: ctx.request.query, variables: ctx.request.variables });
    try {
      const controller = typeof AbortController !== "undefined" ? new AbortController() : null;
      const signal = controller == null ? void 0 : controller.signal;
      let timer;
      if (controller && (opts == null ? void 0 : opts.timeoutMs)) timer = setTimeout(() => controller.abort(), opts.timeoutMs);
      const r = await fetchImpl(ctx.url, {
        method: "POST",
        headers: ctx.headers,
        body,
        signal
      });
      if (timer) clearTimeout(timer);
      const text = await r.text();
      let parsed;
      try {
        parsed = JSON.parse(text);
      } catch (e) {
        resCtx.error = new Error("Invalid JSON response");
        resCtx.status = r.status;
        return;
      }
      resCtx.response = parsed;
      resCtx.status = r.status;
    } catch (err) {
      if (err.name === "AbortError") resCtx.error = new Error("Request timed out");
      else resCtx.error = err;
    }
    await next();
  };
};

// src/client.ts
var GraphQLClient = class {
  async token() {
    return this.tokenProvider();
  }
  async headers() {
    return this.headersFactory();
  }
  getUrl() {
    return this.url;
  }
  constructor(opts) {
    this.url = opts.url;
    this.tokenProvider = toAsyncTokenProvider(opts.tokenProvider);
    this.headersFactory = toAsyncHeadersFactory(opts.headersFactory);
    const base = [this.authMiddleware.bind(this)];
    if (opts.middlewares) base.push(...opts.middlewares);
    base.push(createTransport({ fetchImpl: opts.fetchImpl, timeoutMs: opts.timeoutMs }));
    this.middlewares = base;
  }
  async authMiddleware(ctx, resCtx, next) {
    var _a;
    const token = await this.tokenProvider();
    const dynamicHeaders = await this.headersFactory();
    ctx.headers = {
      "Content-Type": "application/json",
      Accept: "application/json",
      ...dynamicHeaders,
      ...token ? { Authorization: `Bearer ${token}` } : {},
      ...(_a = ctx.headers) != null ? _a : {}
    };
    await next();
  }
  // core request
  async request(query, variables, options) {
    var _a;
    const ctx = {
      request: { query, variables },
      url: this.url,
      headers: (_a = options == null ? void 0 : options.headers) != null ? _a : {},
      cacheOptions: options == null ? void 0 : options.cacheOptions
    };
    const resCtx = {};
    const runner = compose(this.middlewares);
    await runner(ctx, resCtx);
    if (resCtx.error) throw resCtx.error;
    if (!resCtx.response) throw new Error("No response from GraphQL transport");
    if (resCtx.response.errors && resCtx.response.errors.length) {
      console.log("Detailed error - log: ", JSON.stringify(resCtx.response.errors));
      const e = resCtx.response.errors[0];
      const err = new Error(e.message);
      err.extensions = e.extensions;
      throw err;
    }
    return resCtx.response;
  }
};

// src/helpers/entity.factory.ts
function createEntityIntegration(config) {
  var _a;
  return {
    responseFields: [config.key],
    nestedFields: {
      ...(_a = config.nested) != null ? _a : {},
      [config.key]: config.fields
    }
  };
}
function createListIntegration(config) {
  var _a;
  const pluralKey = config.key.endsWith("y") ? `${config.key.slice(0, -1)}ies` : `${config.key}s`;
  return {
    responseFields: [pluralKey, "total"],
    nestedFields: {
      ...(_a = config.nested) != null ? _a : {},
      [pluralKey]: config.fields
    }
  };
}
function createDeleteIntegration(key) {
  const idKey = `${key}Id`;
  return {
    responseFields: [idKey]
  };
}
function createStandardEntityIntegration(config) {
  const base = createEntityIntegration(config);
  return {
    get: base,
    create: base,
    update: base
  };
}

// src/helpers/query.ts
function gqlQueryStringBuilder(query, nestedValues) {
  let queryString = "";
  if (query)
    for (let i = 0; i < query.length; i++) {
      const element = query == null ? void 0 : query[i];
      if (nestedValues == null ? void 0 : nestedValues[element]) {
        queryString += `${String(element)} { 
          ${gqlQueryStringBuilder(nestedValues[element], nestedValues)} 
        }`;
      } else queryString += `${String(element)} `;
    }
  return queryString;
}

// src/helpers/schema-builder.ts
function buildSchema({
  operation,
  name,
  variables,
  field
}) {
  return (selection) => `
    ${operation} ${name}${variables} {
      ${name}${field} {
        ${selection}
      }
    }
  `.trim();
}

// src/helpers/service.factory.ts
function createOperationExecutor2(client, key, config) {
  return async (input, fetchFields, option) => {
    var _a, _b, _c;
    const query = config.schema(
      gqlQueryStringBuilder(
        (_a = fetchFields == null ? void 0 : fetchFields.root) != null ? _a : config.defaultRootFields,
        (_b = fetchFields == null ? void 0 : fetchFields.nestedFields) != null ? _b : config.defaultNestedFields
      )
    );
    const res = await client.request(
      query,
      input,
      option
    );
    return (_c = res.data) == null ? void 0 : _c[key];
  };
}
function createOperationExecutor(client, key, config) {
  return async (input, fetchFields, option) => {
    var _a, _b, _c;
    const query = config.schema(
      gqlQueryStringBuilder(
        (_a = fetchFields == null ? void 0 : fetchFields.root) != null ? _a : config.defaultRootFields,
        (_b = fetchFields == null ? void 0 : fetchFields.nestedFields) != null ? _b : config.defaultNestedFields
      )
    );
    const res = await client.request(
      query,
      input,
      option
    );
    return (_c = res.data) == null ? void 0 : _c[key];
  };
}
function buildService(client, operations) {
  const service = {};
  for (const key in operations) {
    service[key] = createOperationExecutor(
      client,
      key,
      operations[key]
    );
  }
  return service;
}

exports.AuthenticationError = AuthenticationError;
exports.GraphQLClient = GraphQLClient;
exports.NetworkError = NetworkError;
exports.SdkError = SdkError;
exports.buildSchema = buildSchema;
exports.buildService = buildService;
exports.compose = compose;
exports.createCacheMiddleware = createCacheMiddleware;
exports.createDeleteIntegration = createDeleteIntegration;
exports.createEntityIntegration = createEntityIntegration;
exports.createListIntegration = createListIntegration;
exports.createOperationExecutor = createOperationExecutor;
exports.createOperationExecutor2 = createOperationExecutor2;
exports.createStandardEntityIntegration = createStandardEntityIntegration;
exports.createTransport = createTransport;
exports.gqlQueryStringBuilder = gqlQueryStringBuilder;
exports.retryMiddleware = retryMiddleware;
exports.toAsyncHeadersFactory = toAsyncHeadersFactory;
exports.toAsyncTokenProvider = toAsyncTokenProvider;
